File: semester-2/lab-5/task-1/C++/Triangle.cpp
//
// Created by Arthur Ladchenko on 14.04.2020.
//

#include "Triangle.h"
#include <iostream>
#include <cmath>

using namespace std;

Triangle::Triangle(int *coords, int size) : Shape(coords, size) {
    if (size != 6) {
        throw std::runtime_error("Triangle must have 3 and only 3 points.");
    }
}

void Triangle::TriangleData() {
    for (int i = 0; i < this->GetSize(); i++) {
        cout << "Point #";
        cout << i << endl;
        cout << "X: ";
        cout << this->points[i].x << endl;
        cout << "Y: ";
        cout << this->points[i].y << endl;
    }
}

float Triangle::TrianglePerimeter() {
    float perimeter = 0;
    for (int i = 0; i < this->GetSize(); i++) {
        perimeter += this->SideLength(i);
    }

    return perimeter;
}

float Triangle::TriangleArea() {
    float s = this->TrianglePerimeter() / 2;
    float a = this->SideLength(0);
    float b = this->SideLength(1);
    float c = this->SideLength(2);
    return sqrt(s * (s - a) * (s - b) * (s - c));
}

File: semester-2/lab-5/task-1/C#/Triangle.cs
using System;

namespace lab_5
{
    public class Triangle : Shape
    {

        public Triangle(params int[] coords) : base(coords)
        {
            if (coords.Length != 6)
            {
                throw new Exception("Triangle must have 3 and only 3 points.");
            }
        }

        public void TriangleData()
        {
            for (int i = 0; i < Points.Length; i++)
            {
                Console.WriteLine("Point #" + i);
                Console.WriteLine("X: " + Points[i].X + " Y: " + Points[i].Y);
            }
        }
        
        public float TrianglePerimeter()
        {
            float perimeter = 0;
            for (int i = 0; i < Points.Length; i++)
            {
                perimeter += SideLength(i);
            }

            return perimeter;
        }
        
        public float TriangleArea()
        {
            float s = TrianglePerimeter() / 2;
            float a = SideLength(0);
            float b = SideLength(1);
            float c = SideLength(2);
            return (float) Math.Sqrt(s * (s - a) * (s - b) * (s - c));
        }
        
    }
} 